File: Algorithms/164.MaximumGap.java
public class Solution {
    public int maximumGap(int[] nums) {
        int n = nums.length;
        if (n < 2)
            return 0;
        int A = nums[0];
        int B = nums[0];
        for (int i=0; i < n; i++) {
            if (nums[i] > A) 
                A = nums[i];
            if (nums[i] < B) 
                B = nums[i];
        }
        int bucketRange = (int)(((A - B)*1.0 / (n - 1)) + 0.5);
        if (bucketRange < 1)
            bucketRange = 1;
        int bucketLen = (int)(((A - B)*1.0 / bucketRange) + 0.5) + 1;
        int[] buckets = new int[2*bucketLen];
        for (int i = 0; i < 2*bucketLen; i++)
            buckets[i] = -1;
        for (int i = 0; i < n; i++) {
            int loc = (int)(((nums[i] - B)*1.0 / bucketRange) + 0.5);
            if (buckets[loc*2] == -1) {
                buckets[loc*2] = nums[i];
                buckets[loc*2+1] = nums[i];
                continue;
            }
            if (buckets[loc*2] > nums[i])
                buckets[loc*2] = nums[i];
            if (buckets[loc*2+1] < nums[i])
                buckets[loc*2+1] = nums[i];
        }
        int maxGap = 0;
        int i = 0; 
        while (i < bucketLen) {
            if (buckets[2*i] == -1) {
                i ++;
                continue;
            }
            int j = i + 1;
            while (j < bucketLen) {
                if (buckets[2*j] == -1) {
                    j ++;
                    continue;
                }
                if ((buckets[2*j] - buckets[i*2+1]) > maxGap)
                    maxGap = buckets[2*j] - buckets[2*i+1];
                break;
            }
            i = j;
        }
        return maxGap;
    }
}

File: Algorithms/py/164.MaximumGap.py
class Solution:
    # @param nums, a list of integer
    # @return an integer
    def maximumGap(self, nums):
        if len(nums) < 2:
            return 0
        A = max(nums)
        B = min(nums)
        bucketRange = max(1, int(math.ceil((A - B) / (len(nums) - 1))))
        bucketLen = int(math.ceil((A - B) / bucketRange))
        buckets = [None] * (bucketLen + 1)
        for numsber in nums:
            loc = int(math.ceil((numsber - B) / bucketRange))
            bucket = buckets[loc]
            if bucket:
                bucket['min'] = min(bucket['min'], numsber)
                bucket['max'] = max(bucket['max'], numsber)
            else:
                bucket = {'min':numsber, 'max':numsber}
                buckets[loc] = bucket
        maxGap = 0
        for i in range(len(buckets)):
            if not buckets[i]:
                continue
            j = i + 1
            while j < len(buckets):
                if not buckets[j]:
                    j += 1
                    continue
                maxGap = max(maxGap, buckets[j]['min'] - buckets[i]['max'])
                break
        return maxGap
